Add PAR support

josephdecock opened the pull request.

josephdecock committed:
Initial implementation of PAR

josephdecock committed:
Use PAR in sample clients

josephdecock committed:
Minor whitespace correction

josephdecock commented on src/OidcClient/AuthorizeClient.cs:
```diff
@@ -86,7 +85,7 @@ public AuthorizeClient(OidcClientOptions options)
         }
 
-        public AuthorizeState CreateAuthorizeState(Parameters frontChannelParameters)
+        public async Task<AuthorizeState> CreateAuthorizeStateAsync(Parameters frontChannelParameters)
```
We need this breaking change because we are now doing async work here (calling the par endpoint). I think this breaking change results in the best API, but of course we don't like breaking changes.

Options I see:
- Accept the breaking change, since this is a major release
- Keep the old method for backwards compatibility
    - Maybe also mark the old method as deprecated
    - And maybe check if par is enabled when calling the old method. If so, we can't respect the combination of options and the user is clearly already making changes and trying to use the new API, so we can throw an error

brockallen replied:
I'd say make the breaking change, and drop the old method. It's so low level that the chances of people using it directly are low, and we don't document it formally anyhoo.

brockallen replied:
Ping @leastprivilege for any further input.

josephdecock replied:
We discussed and all agree to make this change.

josephdecock commented on src/OidcClient/AuthorizeClient.cs:
```diff
@@ -99,13 +98,46 @@ public AuthorizeState CreateAuthorizeState(Parameters frontChannelParameters)
+                    _logger.LogError("Failed to push authorization parameters");
+
+                    // TODO - Consider how to signal errors to the caller/which exception type to throw
+                    throw new InvalidOperationException(parResponse.Error);
```
@brockallen or @leastprivilege - do you have any suggestion/preference for what the behavior ought to be if the PAR request fails?

josephdecock replied:
Outcome of our discussion: Include errors in the response object instead of throwing, to be consistent with other endpoint calls, e.g., refresh.

josephdecock commented on src/OidcClient/AuthorizeClient.cs:
```diff
@@ -99,13 +98,46 @@ public AuthorizeState CreateAuthorizeState(Parameters frontChannelParameters)
+                ClientId = _options.ClientId,
+                
+                ClientSecret = _options.ClientSecret,
+                ClientAssertion = _options.ClientAssertion,
```
When #422 merges, we should use it here too.

josephdecock committed:
Minor formatting of proj file

josephdecock committed:
Add error message to AuthorizeState

If PAR fails, we now surface that failure in the same way that other response types do.

josephdecock committed:
Clean up test clients

- Remove defunct ConsoleClient
- Add a manual mode console client

josephdecock committed:
Enable PAR by default when it is supported by the provider

josephdecock committed:
Fix up solution file

josephdecock committed:
Add github readme to solution

josephdecock committed:
Merge branch 'main' into joe/par

josephdecock committed:
Use GetClientAssertionAsync in PAR

josephdecock committed:
Suppress strong name warning in dpop test proj

We don't care about strong naming in general, and we definitely don't care in a test project

brockallen commented on src/OidcClient/AuthorizeClient.cs:
```diff
@@ -98,16 +105,18 @@ public async Task<AuthorizeState> CreateAuthorizeStateAsync(Parameters frontChan
-                    throw new InvalidOperationException(parResponse.Error);
+                    state.Error = "Failed to push authorization parameters";
+                    state.ErrorDescription = parResponse.Error;
+                    return state;
```
Hmm... should we keep the error as the error on the result object?

josephdecock committed:
Swap error and error description

brockallen commented on src/OidcClient/AuthorizeClient.cs:
```diff
@@ -114,8 +114,8 @@ public async Task<AuthorizeState> CreateAuthorizeStateAsync(Parameters frontChan
-                    state.Error = "Failed to push authorization parameters";
-                    state.ErrorDescription = parResponse.Error;
+                    state.Error = parResponse.Error;
+                    state.ErrorDescription = "Failed to push authorization parameters";
```
Yes, and maybe also:

`state.ErrorDescription = "Failed to push authorization parameters" + parResponse.ErrorDescription;`

of course with some better formatting and null checking, etc

josephdecock replied:
The par response only has a single error property, inherited from the base class in IdentityModel
